### PR TITLE
Adding QueryDocumentSnapshot

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
This is based on user feedback and changes the DocumentSnapshot API to return `null` for non-existing documents (instead of throwing an Exception). Since Queries never return `null` documents, this PR also introduces `QueryDocumentSnapshot`, which overrides the `DocumentSnapshot` methods to be non-nullable.

Addresses b/71489271 and b/69365491

### CODE REVIEW

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/DocumentSnapshot.java [hunk 1]
```diff
@@ -34,9 +34,13 @@
 
 /**
  * A DocumentSnapshot contains data read from a document in a Firestore database. The data can be
- * extracted with the getData or get methods.
+ * extracted with the getData() or get() methods.
```
Comment (pongad) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/DocumentSnapshot.java]: `{@code getData()} or {@code get()}`?

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/DocumentSnapshot.java [hunk 2]
```diff
@@ -189,51 +194,54 @@ public boolean exists() {
...
+   * @return The fields of the document as a Map or null if the document doesn't exist.
    */
-  @Nonnull
+  @Nullable
   public Map<String, Object> getData() {
-    Preconditions.checkState(exists(), "Cannot call getData() on missing document.");
-    Map<String, Object> decodedFields = new HashMap<>();
-    for (Map.Entry<String, Value> entry : fields.entrySet()) {
-      decodedFields.put(entry.getKey(), decodeValue(entry.getValue()));
+    if (fields != null) {
```
Comment (pongad) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/DocumentSnapshot.java]: I prefer `if (fields == null) return null` to reduce nesting, but your choice.

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/QueryDocumentSnapshot.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+   * Returns the fields of the document as a Map. Field values will be converted to their native
+   * Java representation.
+   *
+   * @return The fields of the document as a Map.
+   */
+  @Nonnull
+  @Override
+  public Map<String, Object> getData() {
+    Map<String, Object> result = super.getData();
+    Preconditions.checkNotNull(result, "Data in a QueryDocumentSnapshot should be non-null");
```
Comment (pongad) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/QueryDocumentSnapshot.java]: You can inline this: `return checkNotNull(super.getData(), "...");`
I slightly prefer the inline version, but if you think the "expanded form" reads better, I'm good with this as is.

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/QueryDocumentSnapshot.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+   * Returns the contents of the document converted to a POJO.
+   *
+   * @param valueType The Java class to create
+   * @return The contents of the document in an object of type T
+   */
+  @Nonnull
+  @Override
+  public <T> T toObject(@Nonnull Class<T> valueType) {
+    T result = super.toObject(valueType);
+    Preconditions.checkNotNull(result, "Object in a QueryDocumentSnapshot should be non-null");
```
Comment (pongad) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/QueryDocumentSnapshot.java]: Same comment as above.

**Comment by pongad** [comment]
Some stylistic comments. @garrettjonesgoogle PTAL

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/DocumentSnapshot.java (hunk 1 above)
Comment (schmidt-sebastian) [reply google-cloud-firestore/src/main/java/com/google/cloud/firestore/DocumentSnapshot.java]: Done.

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/DocumentSnapshot.java (hunk 2 above)
Comment (schmidt-sebastian) [reply google-cloud-firestore/src/main/java/com/google/cloud/firestore/DocumentSnapshot.java]: Works for me.

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/QueryDocumentSnapshot.java (hunk 3 above)
Comment (schmidt-sebastian) [reply google-cloud-firestore/src/main/java/com/google/cloud/firestore/QueryDocumentSnapshot.java]: I always forget that these methods can be used like that. In this case, I think it is more readable the way it is though and I kept it as is.

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/QueryDocumentSnapshot.java (hunk 4 above)
Comment (schmidt-sebastian) [reply google-cloud-firestore/src/main/java/com/google/cloud/firestore/QueryDocumentSnapshot.java]: Kept as is.

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/DocumentSnapshot.java [hunk 5]
```diff
@@ -34,9 +34,13 @@
 
 /**
  * A DocumentSnapshot contains data read from a document in a Firestore database. The data can be
- * extracted with the getData or get methods.
+ * extracted with the {@code getData()} or {@code get()} methods.
```
Comment (garrettjonesgoogle) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/DocumentSnapshot.java]: Why not use links for references to methods? e.g. `{@link #getData()}`

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/QueryDocumentSnapshot.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2017 Google Inc. All Rights Reserved.
```
Comment (garrettjonesgoogle) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/QueryDocumentSnapshot.java]: 2018 :-)

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/QueryDocumentSnapshot.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+import com.google.common.base.Preconditions;
+import com.google.firestore.v1beta1.Document;
+import com.google.firestore.v1beta1.Value;
+import com.google.protobuf.Timestamp;
+import java.util.Map;
+import javax.annotation.Nonnull;
+import org.threeten.bp.Instant;
+
+/**
+ * A QueryDocumentSnapshot contains data read from a document in your Firestore database as part of
```
Comment (garrettjonesgoogle) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/QueryDocumentSnapshot.java]: s/in your Firestore database/in a Firestore database/

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/DocumentSnapshot.java (hunk 5 above)
Comment (schmidt-sebastian) [reply google-cloud-firestore/src/main/java/com/google/cloud/firestore/DocumentSnapshot.java]: Done

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/QueryDocumentSnapshot.java (hunk 6 above)
Comment (schmidt-sebastian) [reply google-cloud-firestore/src/main/java/com/google/cloud/firestore/QueryDocumentSnapshot.java]: Done

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/QueryDocumentSnapshot.java (hunk 7 above)
Comment (schmidt-sebastian) [reply google-cloud-firestore/src/main/java/com/google/cloud/firestore/QueryDocumentSnapshot.java]: Done

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/QueryDocumentSnapshot.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2018 Google Inc. All Rights Reserved.
```
Comment (garrettjonesgoogle) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/QueryDocumentSnapshot.java]: I know this is a pain, but could you update the copyright line? We updated it across the codebase. It should be:

```
Copyright 2018 Google LLC
```

Comment (schmidt-sebastian) [reply google-cloud-firestore/src/main/java/com/google/cloud/firestore/QueryDocumentSnapshot.java]: No problem, done.

**Pull request merged**